### ISSUE TITLE
The 'exists' function needs expansion, not depreciation

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11504]
Migrated from [rt.perl.org#94972](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=94972) (status was 'rejected')

Searchable as RT94972$

**Comment by p5pRT** [comment]
### From sineswiper@gmail.com

Traditionally\, Perl has maintained their version of the NULL value = undef\.
This aspect of variables needs to be expanded on\, but potentially new
changes may take a step backwards&#8203;:

exists may also be called on array elements\, but its behavior is much less
obvious and is strongly tied to the use of
delete\<http&#8203;://perldoc\.perl\.org/functions/delete\.html>on arrays\.
\* Be aware that calling exists on array values is deprecated and likely to
be removed in a future version of Perl\.\*

There's a need for variables to exist in a state of "initialized\, but
undefined"\.  All variables\, not just hashes\.  And these states need to be
detectable as "initialized\, but undefined"\.  As in "my container's value is
undefined\, but my container still exists\."  Zero\-like values like '' or 0
are different than NULL or undef\.

Instead of removing functionality from the 'exists' function\, it should be
expanded to work on scalars\, arrays\, hashes\, code containers\, references\,
even globs and filehandles\.  If the behavior of undef doesn't work in all
cases\, then that should be fixed as well&#8203;:

$aaa = undef;  \# initialized\, but undefined
undef $bbb;  \# removes the container\, so the variable is now uninitialized
$ccc\[4\] = undef;  \# initialized\, but undefined \(even if it's the last one in
the array\)
$ddd\{'key'\} = undef;  \# initialized\, but undefined
$ddd\{undef\} = undef;  \# still valid
$ccc\[undef\] = undef;  \# illegal \(undef isn't a number\)
undef $ccc\[4\];  \# either illegal or works like splice
undef $ddd\{'key'\};  \# either illegal or works like delete

exists $aaa && \!defined $aaa;  \# true for both
exists $ccc\[4\] && \!defined $ccc\[4\];  \# true for both
exists $ddd\{'key'\} && \!defined $ddd\{'key'\};  \# true for both

In other words\, the left side and the right side of undef should work
differently\.  Making something undefined and undefining something are two
different things\.

\-\- 
Brendan Byrd/SineSwiper \<SineSwiper@&#8203;GMail\.com>
Computer tech\, PERL wizard\, and all\-around Internet guru

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Jul 17 11&#8203;:05&#8203;:37 2011\, sineswiper@&#8203;gmail\.com wrote&#8203;:
> undef $ddd\{'key'\};  \# either illegal or works like delete

I don’t think changing it to work like delete is a good idea\, for
compatibility reasons\.

However\, being able to say ‘delete $foo’ and ‘exists $foo’ sounds like a
very nice idea\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From j.imrie1@virginmedia.com

On 17/07/2011 23&#8203;:04\, Father Chrysostomos via RT wrote&#8203;:
> On Sun Jul 17 11&#8203;:05&#8203;:37 2011\, sineswiper@&#8203;gmail\.com wrote&#8203;:
>> undef $ddd\{'key'\};  \# either illegal or works like delete
> I don’t think changing it to work like delete is a good idea\, for
> compatibility reasons\.
>
> However\, being able to say ‘delete $foo’ and ‘exists $foo’ sounds like a
> very nice idea\.
>
Doesn't delete $foo and exists $foo map to

delete $&#8203;::\{foo\} and exists $&#8203;::\{foo\}

ie manipulations on the stash or it's equivalent for lexical variables\. 
In which case\, are all we are really talking about some syntactic sugar 
to allow stash manipulation/inspection?

John

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Brendan Byrd wrote&#8203;:
>exists may also be called on array elements\, but its behavior is much less
>obvious and is strongly tied to the use of
>delete\<http&#8203;://perldoc\.perl\.org/functions/delete\.html>on arrays\.

delete doesn't conceptually make sense on arrays\, since the deleted
element doesn't stop existing\, and it was a mistake to ever expose the
implementation detail that exists on arrays made visible\.

>There's a need for variables to exist in a state of "initialized\, but
>undefined"\.

You're asking us to introduce a new kind of undef value\, and a new state
for non\-scalar objects\.  This is bad from a backward compatibility point
of view\, and you also haven't made a case for why there is a need for
these new kinds of value\.

\-zefram

**Comment by p5pRT** [comment]
### From BitCard@ResonatorSoft.org

On Mon Jul 18 05&#8203;:06&#8203;:51 2011\, zefram@&#8203;fysh\.org wrote&#8203;:
> Brendan Byrd wrote&#8203;:
>
> delete doesn't conceptually make sense on arrays\, since the deleted
> element doesn't stop existing\, and it was a mistake to ever expose the
> implementation detail that exists on arrays made visible\.

I'm more worried about the functionality for undef and exists\.
 
> You're asking us to introduce a new kind of undef value\, and a new state
> for non\-scalar objects\.

Really?  I'm just asking for an expansion of the existing undef value\. 
Or rather\, just an expansion of the DETECTION of the undef value\.

What is undef?  What does \($abc\{'jkl'\} = undef\) really mean?  How is
that any different from \(sqlVar = NULL\) in SQL?

> This is bad from a backward compatibility point
> of view\, and you also haven't made a case for why there is a need for
> these new kinds of value\.

Because NULL exists for every language\, and we need to interact with
these languages on a daily basis&#8203;:

SQL = NULL
XML = nil
JSON = NULL
PHP = NULL
Python = None
Ruby = nil
Java = null
Java JDBC = wasNull\(\)
JavaScript = null / undefined / NaN
C = NULL
C\+\+ = NULL / void
C\# = null
Objective\-C = nil
Smalltalk = nil / UndefinedObject
Lisp = nil
EmacsLisp = nil / \#f \(\)
XPath = NaN
YAML = \(empty\) / ~ / null
Io = nil
SML = NONE
Haskell = Nothing
Mathematica = NULL
Rebol = none
Eiffel = Void
Lua = nil
Maple = NULL
Mathematics = ∅

Perl = undef

Converting these values to strings like 'NULL' or 'NIL' does not work\,
because a string exists and NULL is the absence of data\.  \(Or rather\,
the confirmation that the data does not exist\.\)  Besides\, it's a merging
of the real string value 'NULL'\, so how can you tell the difference
between real 'NULL' and this fake 'NULL'?

Undef is the ONLY value that uniquely identifies NULL\, or the absence of
data\.  We need to detect it and define it in every shape and form\.

Right now\, how can you tell the difference between&#8203;:

my $aaa = undef;

exists\($aaa\);      \# initialized but undefined
exists\($dsrfgkl\);  \# doesn't even exist

**Comment by p5pRT** [comment]
### From BitCard@ResonatorSoft.org

Oh\, and let's not forget\.\.\.

VB\.NET = Nothing
VB6 = Nothing
Delphi = nil
Foxpro = NULL
Ada = not X'valid
ALGOL 68 = NIL / VOID
D = void
E = null
GAP = IsBound\(X\)
Go = nil
MUMPS = UNDEF
GP = 'X
PARI = 'X / is\_entry\("X"\)
R = NULL
Perl 6 = Nil \(yes\, Perl 6\)

**Comment by p5pRT** [comment]
### From @abigail

On Mon\, Jul 18\, 2011 at 01&#8203;:06&#8203;:17PM \+0100\, Zefram wrote&#8203;:
> Brendan Byrd wrote&#8203;:
> >exists may also be called on array elements\, but its behavior is much less
> >obvious and is strongly tied to the use of
> >delete\<http&#8203;://perldoc\.perl\.org/functions/delete\.html>on arrays\.
> 
> delete doesn't conceptually make sense on arrays\, since the deleted
> element doesn't stop existing\, and it was a mistake to ever expose the
> implementation detail that exists on arrays made visible\.

I'd argue that delete conceptually makes a lot of sense \- but I agree
that it's original implementation neither made much sense\, nor was 
very useful\.

I think that

&nbsp;    delete @&#8203;array\[@&#8203;list\]

being a short hand for

&nbsp;    use List&#8203;::MoreUtils 'uniq';
&nbsp;    foreach my $index \(sort \{$b \<=> $a\} uniq @&#8203;list\) \{
&nbsp;        splice @&#8203;array\, $index\, 1;
&nbsp;    \}

or

&nbsp;    @&#8203;array = do \{my %h = map \{$\_ => 1\} @&#8203;list;
&nbsp;                 @&#8203;array \[grep \{\!$h \{$\_\}\} 0 \.\. $\#array\]\};

makes a lot of sense\.



Abigail

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Jul 21 03&#8203;:44&#8203;:09 2011\, abigail@&#8203;abigail\.be wrote&#8203;:
> On Mon\, Jul 18\, 2011 at 01&#8203;:06&#8203;:17PM \+0100\, Zefram wrote&#8203;:
> > Brendan Byrd wrote&#8203;:
> > >exists may also be called on array elements\, but its behavior is
> much less
> > >obvious and is strongly tied to the use of
> > >delete\<http&#8203;://perldoc\.perl\.org/functions/delete\.html>on arrays\.
> >
> > delete doesn't conceptually make sense on arrays\, since the deleted
> > element doesn't stop existing\, and it was a mistake to ever expose
> the
> > implementation detail that exists on arrays made visible\.
> 
> I'd argue that delete conceptually makes a lot of sense \- but I agree
> that it's original implementation neither made much sense\, nor was
> very useful\.

It is useful\. I use it all the time\. It works well in conjunction with
Class&#8203;::ArrayObjects\.

**Comment by p5pRT** [comment]
### From @doy

Unless I'm mistaken\, I think this is a "wontfix"\.

\-doy

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Unless I'm mistaken\, I think this is a "wontfix"\.

\-doy

**Comment by p5pRT** [comment]
@doy - Status changed from 'open' to 'rejected'
